tracing-subscriber/src/layer.rs

```diff
@@ -0,0 +1,255 @@
+    }
+    fn new_span(&self, _attrs: &span::Attributes, _id: &span::Id) {}
+    fn record(&self, _span: &span::Id, _values: &span::Record) {}
+    fn record_follows_from(&self, _span: &span::Id, _follows: &span::Id) {}
+    fn event(&self, _event: &Event) {}
+    fn enter(&self, _id: &span::Id) {}
+    fn exit(&self, _id: &span::Id) {}
+    fn clone_span(&self, _id: &span::Id, _new: Option<&span::Id>) {}
+    fn drop_span(&self, _id: &span::Id) {}
+
+    /// Composes the given [`Subscriber`] with this `Layer`, returning a `Layered` subscriber.
+    ///
+    /// The returned `Layered` subscriber will call the methods on this `Layer`
+    /// and then those of the wrapped subscriber. Multiple layers may be
+    /// composed in this manner. For example:
+    /// ```rust
+    /// # use tracing_subscriber::layer::Layer;
+    /// # fn main() {
+    /// pub struct FooLayer {
+    ///     // ...
+    /// }
+    ///
+    // /// pub struct BarLayer {
+    // ///     // ...
+    // /// }
+    // ///
+    /// pub struct MySubscriber {
+    ///     // ...
+    /// }
+    ///
+    /// impl<S> Layer<S> for FooLayer {
+    ///     // ...
+    /// }
+    ///
+    /// # impl FooLayer {
+    /// # fn new() -> Self { Self {} }
+    /// # }
+    /// # impl MySubscriber {
+    /// # fn new() -> Self { Self { }}
+    /// # }
+    /// # use tracing_core::{span::{Id, Attributes, Record}, Metadata};
+    /// # impl tracing_core::Subscriber for MySubscriber {
+    /// #   fn new_span(&self, _: &Attributes) -> Id { Id::from_u64(0) }
+    /// #   fn record(&self, _: &Id, _: &Record) {}
+    /// #   fn event(&self, _: &tracing_core::Event) {}
+    /// #   fn record_follows_from(&self, _: &Id, _: &Id) {}
+    /// #   fn enabled(&self, _: &Metadata) -> bool { false }
+    /// #   fn enter(&self, _: &Id) {}
+    /// #   fn exit(&self, _: &Id) {}
+    /// # }
+    /// let subscriber = FooLayer::new()
+    // ///     .and_then(BarLayer::new()
+    ///     .and_then(MySubscriber::new());
+    /// # }
+    fn and_then(self, inner: S) -> Layered<Self, S>
+    where
+        Self: Sized,
+    {
+        Layered { layer: self, inner }
+    }
+}
+
+pub trait SubscriberExt: Subscriber + crate::sealed::Sealed {
+    fn with<L>(self, layer: L) -> Layered<L, Self>
+    where
+        L: Layer<Self>,
+        Self: Sized,
+    {
+        Layered { layer, inner: self }
+    }
+
+    // fn with_enabled<F>(self, f: F) -> Layered<filter::EnabledFn<F>, Self>
+    // where
+    //     F: Fn(&Metadata) -> bool + 'static,
+    //     Self: Sized,
+    // {
+    //     self.layer(filter::enabled_fn(f))
+    // }
+
+    // fn with_callsite_filter<F>(self, f: F) -> Layered<filter::InterestFn<F>, Self>
+    // where
+    //     F: Fn(&Metadata) -> Interest + 'static,
+    //     Self: Sized,
+    // {
+    //     self.layer(filter::enabled_fn(f))
+    // }
+}
+
+#[derive(Clone, Debug)]
+pub struct Layered<L, S> {
+    layer: L,
+    inner: S,
+}
+
+// === impl Layered ===
+
+impl<A, B> Layered<A, B> {
```

Comment:
Assuming a tree structure of subscribers, is `Layered<A, B>` meant to provide a mechanism to "vertically" compose subscribers, while `And`/`Or` is meant to provide "horizontal" composition of subscribers (e.g., within the same level in the tree structure)?

Reply:
`And`/`Or` is for composing _filters_ via boolean operations to make a new filter. Layer is intended for composing event consumers/filters.

Reply:
This branch doesn't implement a tree structure of subscribers, just vertical composition. I'm not sure if a tree structure really makes sense or is necessary.